packages/core/saved-objects/core-saved-objects-migration-server-internal/src/document_migrator/pipelines/downgrade_pipeline.ts

```diff
@@ -0,0 +1,137 @@
+import type { MigrationPipeline, MigrationPipelineResult } from './types';
+import { applyVersion, assertValidCoreVersion } from './utils';
+
+export class DocumentDowngradePipeline implements MigrationPipeline {
+  private document: SavedObjectUnsanitizedDoc;
+  private kibanaVersion: string;
+  private originalDoc: SavedObjectUnsanitizedDoc;
+  private typeTransforms: TypeTransforms;
+  private targetTypeVersion: string;
+  private targetCoreVersion?: string;
+
+  constructor({
+    document,
+    kibanaVersion,
+    typeTransforms,
+    targetTypeVersion,
+    targetCoreVersion,
+  }: {
+    document: SavedObjectUnsanitizedDoc;
+    typeTransforms: TypeTransforms;
+    kibanaVersion: string;
+    targetTypeVersion: string;
+    targetCoreVersion?: string;
+  }) {
+    this.originalDoc = document;
+    this.kibanaVersion = kibanaVersion;
+    this.document = cloneDeep(document);
+    this.typeTransforms = typeTransforms;
+    this.targetTypeVersion = targetTypeVersion;
+    this.targetCoreVersion = targetCoreVersion;
+  }
+
+  run(): MigrationPipelineResult {
+    assertValidCoreVersion({ document: this.document, kibanaVersion: this.kibanaVersion });
+    this.assertCompatibility();
+
+    for (const transform of this.getPendingTransforms()) {
+      if (!transform.transformDown) {
+        throw new Error(
+          `Could not apply transformation ${transform.transformType}:${transform.version}: no down conversion registered`
+        );
+      }
+      const { transformedDoc } = transform.transformDown(this.document);
+      if (this.document.type !== this.originalDoc.type) {
+        throw new Error(`Changing a document's type during a migration is not supported.`);
+      }
+      this.document = applyVersion({ document: transformedDoc, transform });
+    }
+
+    this.document = this.ensureVersion(this.document);
+
+    return {
+      document: this.document,
+      additionalDocs: [],
+    };
+  }
+
+  private getPendingTransforms() {
+    const { transforms } = this.typeTransforms;
+    return transforms.reverse().filter((transform) => this.isPendingTransform(transform));
+  }
+
+  private isPendingTransform({ transformType, version }: Transform) {
+    const { coreMigrationVersion, typeMigrationVersion } = this.document;
+
+    switch (transformType) {
+      // reference and convert type transforms were deprecated before downward conversion were a thing
+      // so we will never need to revert such type of transforms (and they didn't implement a down conversion)
+      case TransformType.Reference:
+      case TransformType.Convert:
+        return false;
+      // including core transforms if targetCoreVersion is specified
+      case TransformType.Core:
+        return (
+          this.targetCoreVersion &&
+          (coreMigrationVersion == null ||
+            Semver.gt(coreMigrationVersion, this.targetCoreVersion)) &&
+          Semver.gt(version, this.targetCoreVersion)
+        );
```

Comment:
I implemented support for downward `core` transforms, even if it's not planned to use it atm.